share/p2p/peers/manager.go

```diff
@@ -0,0 +1,234 @@
+	case <-ctx.Done():
+		return "", nil, ctx.Err()
+	}
+}
+
+func (s *Manager) doneFunc(datahash share.DataHash, peerID peer.ID) DoneFunc {
+	return func(result sampleResult) {
+		switch result {
+		case ResultSuccess:
+			s.markSampled(datahash)
+		case ResultFail:
+		case ResultPeerMMisbehaved:
+			// TODO: signal to peers Validator to return Reject
+			s.RemovePeers(datahash, peerID)
+			s.fullNodes.remove(peerID)
+		}
+	}
+}
+
+// markSampled marks datahash as sampled if not yet marked, to release waiting validator and
+// retransmit the message via shrex.Sub
+func (s *Manager) markSampled(datahash share.DataHash) {
+	p := s.getOrCreateValidatedPool(datahash.String())
+	if p.isSynced.CompareAndSwap(false, true) {
+		close(p.waitSyncCh)
+	}
+}
+
+// RemovePeers removes peers for given datahash from store
+func (s *Manager) RemovePeers(datahash share.DataHash, ids ...peer.ID) {
+	p := s.getOrCreateValidatedPool(datahash.String())
+	p.pool.remove(ids...)
+}
+
+// subscribeHeader marks pool as validated when its datahash corresponds to a header received from
+// headerSub.
+func (s *Manager) subscribeHeader(ctx context.Context) {
+	defer close(s.done)
+	defer s.headerSub.Cancel()
+
+	for {
+		h, err := s.headerSub.NextHeader(ctx)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return
+			}
+			log.Errorw("get next header from sub", "err", err)
+			continue
+		}
+
+		s.getOrCreateValidatedPool(h.DataHash.String())
+	}
+}
+
+func (s *Manager) getOrCreateValidatedPool(datahash string) syncPool {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	p, ok := s.pools[datahash]
+	if !ok {
+		p = newSyncPool()
+
+		// save as already validated
+		p.isValidDataHash.Store(true)
+		s.pools[datahash] = p
+		return p
+	}
+
+	// if not yet validated, there is validator waiting that needs to be released
+	p.markValidated()
+	return p
+}
+
+// Validate will block until header with given datahash received. This behavior opens an attack
+// vector of multiple fake datahash spam, that will grow amount of hanging routines in node. To
+// address this, validator should be reworked to be non-blocking, with retransmission being invoked
+// in sync manner from another routine upon header discovery.
+func (s *Manager) Validate(ctx context.Context, peerID peer.ID, hash share.DataHash) pubsub.ValidationResult {
+	p := s.getOrCreateUnvalidatedPool(hash.String())
+	p.pool.add(peerID)
+
+	// check of validation required
+	if !p.isValidDataHash.Load() {
+		if valid := p.waitValidation(ctx); !valid {
+			// no corresponding header was received for given datahash in time,
+			// highly unlikely block with given datahash exist in chain, reject msg and punish the peer
+			s.deletePool(hash.String())
+			return pubsub.ValidationReject
+		}
+	}
+
+	// headerSub found corresponding ExtendedHeader for dataHash,
+	// wait for it to be sampled, before retransmission
+	if sampled := p.waitSampling(ctx); sampled {
+		return pubsub.ValidationAccept
+	}
```

Comment:
doc nit: The use of sampling terminology here is not correct. Sampling is a concern of dasing; here, we only ensure the data is loaded/synced/fetched in it's entirety, even though we all understand what we mean here by sampling.